wordpress-mobile/WordPress-Android
# Adds WPUtils as composite build

This PR takes over the excellent work @oguzkocer started in https://github.com/wordpress-mobile/WordPress-Android/pull/12796. 
The PR was opened some time ago and there are now conflicts with the Utils library. I tried to fix those, but due to the involved operations, the graph of the branch ended up to be very confusing, so I preferred to branch from https://github.com/wordpress-mobile/WordPress-Android/pull/12796/commits/f19af8bdba0f16a980c8179ab338a18aec377e01 which contains all the diffs to enable the composite build and re-do the deletion of the current `Utils` library in this branch. 

The changes in this PR can be split in three parts: 

1. changes required to enable the composite build. An easy way to review them is to look at this branch: https://github.com/wordpress-mobile/WordPress-Android/pull/13465 which is cut off https://github.com/wordpress-mobile/WordPress-Android/pull/12796/commits/f19af8bdba0f16a980c8179ab338a18aec377e01 (I'm keeping it open just to this review easier. I'll delete it without merging it). 

2. changes to remove the `utils` library subtree. They are done in the commits from https://github.com/wordpress-mobile/WordPress-Android/pull/13466/commits/b4490675f24db19c073425936ba485e1c194766f to https://github.com/wordpress-mobile/WordPress-Android/pull/13466/commits/45a13efb5bbf1a7a48e4dc50a18611fdd3fc50be.

3. changes to make lint work without errors after that. 


**To test:**
For testing the first two points above I'll just copy Oguz's instructions: 

This PR is a follow up to #13001 and makes https://github.com/wordpress-mobile/WordPress-Utils-Android a binary dependency and adds the option to make it a composite build for local development. It also sets up the repository to remove the libs/utils subtree since it's no longer used.

Please make sure to check out #13001 for details on what composite build is and how it works. Since the merge of this PR will be a bit involved, I am not duplicating that information here to keep things brief. Also note that, the PR currently includes changes from #13001 since it's not yet merged, but once it is merged, the changes should stop showing up. (sometimes it doesn't in which case, please change the base branch to another branch and change it back to develop which hopefully resolves the issue)

**To test:**

- Copy local-builds.gradle-example as local-builds.gradle and update localWPUtilsPath to your liking
- Use Navigate Class shortcut to open UrlUtils which is in WPUtils and add a new static method such as a summation
- Use Navigate Class shortcut to open ReaderPostLogic (any other class would be fine) which is in WPAndroid and use that function
- Run the project and verify that it works 🎉
- Comment out localWPUtilsPath in local-builds.gradle file and verify that the build fails for missing the new method from UrlUtils
- Discard the changes for ReaderPostLogic, build the app and verify that it works

**Lint errors**

I found a few lint errors after removing the subtree. I dug into those, but I couldn't find a lot of information about them. They are mostly about missing classes and wrong interfaces, for which I expected crashes if they were actual issues, but I've tested the app and I haven't experienced any crash, so I _think_ that they are mostly related to the linter not handling well composite builds, **but** I may be very wrong here, so, please, lay out any doubt you may have. 

1. The first error is here: https://github.com/wordpress-mobile/WordPress-Android/pull/13466/commits/e911e43f196bc2d54d8511e70de88e6a1499dead
This happens when running `gradlew lintVanillaRelease` in binary import mode. I downloaded `aar` file, decompiled it and verified that the parameter expects a `ColorInt`, so it looked good to me. So, I silenced it as a false positive.

2. The second group of errors is here: https://github.com/wordpress-mobile/WordPress-Android/pull/13466/commits/af3cbeff9b0d7d34ae313c77111350e88aac08ba. They happen when running composite builds with dependency substitution in local mode. I expected build errors if classes where really missing, but the app builds fine and testing it it seems to work ok. So, I silenced them as false positives. 

3. The third error is here: https://github.com/wordpress-mobile/WordPress-Android/pull/13466/commits/b70fed75cc64a89a05a3910b2cb243b8250f27ca. This is the only one I'm sure about why it happens as it's because we import the resource in our main `strings.xml` to have it translated. The fix for this one should be ok. 

4. The last group of errors is here: https://github.com/wordpress-mobile/WordPress-Android/pull/13466/commits/57449da95c0d685d97ea4c5191ef80e3e5b49352. Again, I think that it's because the lint doesn't compute the dependencies right with composite builds, but I'm not totally sure about why. I silenced them as false positives as the app seems to work well. 


PR submission checklist:
